File: Python/student_grade.py
def main():
    print("How many students do you have?")
    num_students = get_positive_input("Enter a positive integer: ")
    
    print("\nHow many subjects do they offer?")
    num_subjects = get_positive_input("Enter a positive integer: ")
    print("Saved successfully\n")
    
    names, grades = collect_student_data(num_students, num_subjects)
    totals = calculate_totals(grades)
    averages = calculate_averages(totals, num_subjects)
    positions = calculate_positions(totals)
    
    print_report_header(num_students, num_subjects)
    print_student_results(names, grades, totals, averages, positions)
    print_subject_summaries(names, grades)
    print_class_statistics(names, grades, totals)


def get_positive_input(prompt):
    while True:
        try:
            value = int(input(prompt))
            if value > 0:
                return value
            print("Invalid input. Please enter a positive integer.")
        except ValueError:
            print("Invalid input. Please enter a positive integer.")

def get_valid_score(prompt):
    while True:
        try:
            score = int(input(prompt))
            if 0 <= score <= 100:
                return score
            print("Invalid score! Enter a value between 0-100.")
        except ValueError:
            print("Invalid input. Please enter a number.")

def collect_student_data(num_students, num_subjects):
    student_names = []
    student_grades = []
    
    for i in range(num_students):
        name = input(f"Enter name for student {i+1}: ")
        student_names.append(name)
        
        grades = []
        for j in range(num_subjects):
            score = get_valid_score(
                f"Enter score for {name} in subject {j+1} (0-100): "
            )
            grades.append(score)
        student_grades.append(grades)
        
    return student_names, student_grades

def calculate_totals(student_grades):
    return [sum(grades) for grades in student_grades]

def calculate_averages(totals, num_subjects):
    return [total / num_subjects for total in totals]

def calculate_positions(totals):
    ranked = sorted(((total, i) for i, total in enumerate(totals)), reverse=True)
    positions = [0] * len(totals)
    current_rank = 1
    
    for i, (total, idx) in enumerate(ranked):
        if i > 0 and total < ranked[i-1][0]:
            current_rank = i + 1
        positions[idx] = current_rank
        
    return positions

def print_report_header(num_students, num_subjects):
    print("\nClass Summary Report")
    print("====================")
    print(f"Number of Students: {num_students}")
    print(f"Number of Subjects: {num_subjects}\n")
    
    header = "Student Name\t\t"
    header += "\t".join(f"Subject {i+1}" for i in range(num_subjects))
    header += "\tTotal\tAverage\tPosition"
    print(header)

def print_student_results(names, grades, totals, averages, positions):
    for i in range(len(names)):
        row = f"{names[i]}\t\t"
        row += "\t".join(str(grade) for grade in grades[i])
        row += f"\t{totals[i]}\t{averages[i]:.2f}\t{positions[i]}"
        print(row)

def print_subject_summaries(names, grades):
    print("\nSubject Summary")
    print("===============")
    
    num_subjects = len(grades[0])
    num_students = len(grades)
    
    for j in range(num_subjects):
        highest = -1
        lowest = 101
        high_name = ""
        low_name = ""
        total = 0
        passes = 0
        
        for i in range(num_students):
            score = grades[i][j]
            total += score
            if score > highest:
                highest = score
                high_name = names[i]
            if score < lowest:
                lowest = score
                low_name = names[i]
            if score >= 50:
                passes += 1
        
        avg = total / num_students
        fails = num_students - passes
        
        print(f"\nSubject {j+1}")
        print(f"Highest scoring student is: {high_name} ({highest})")
        print(f"Lowest scoring student is: {low_name} ({lowest})")
        print(f"Total score: {total}")
        print(f"Average score: {avg:.2f}")
        print(f"Number of passes: {passes}")
        print(f"Number of fails: {fails}")
        print("---------------------")

def print_class_statistics(names, grades, totals):
    print("\nClass Statistics")
    print("================")
    
    # Hardest/easiest subjects
    num_subjects = len(grades[0])
    num_students = len(grades)
    
    max_fails = -1
    min_fails = num_students + 1
    hardest = -1
    easiest = -1
    
    for j in range(num_subjects):
        fails = sum(1 for i in range(num_students) if grades[i][j] < 50)
        if fails > max_fails:
            max_fails = fails
            hardest = j
        if fails < min_fails:
            min_fails = fails
            easiest = j
            
    print(f"The hardest subject is Subject {hardest+1} with {max_fails} failures")
    print(f"The easiest subject is Subject {easiest+1} with {num_students-min_fails} passes")
    
    # Overall highest/lowest
    high_score = -1
    low_score = 101
    high_name = ""
    low_name = ""
    high_subj = -1
    low_subj = -1
    
    for i in range(num_students):
        for j in range(num_subjects):
            score = grades[i][j]
            if score > high_score:
                high_score = score
                high_name = names[i]
                high_subj = j
            if score < low_score:
                low_score = score
                low_name = names[i]
                low_subj = j
                
    print(f"The overall highest score is scored by {high_name} in Subject {high_subj+1} ({high_score})")
    print(f"The overall lowest score is scored by {low_name} in Subject {low_subj+1} ({low_score})")
    
    # Class summary
    print("\nClass Summary")
    print("=============")
    
    best_idx = totals.index(max(totals))
    worst_idx = totals.index(min(totals))
    
    print(f"Best graduating student is {names[best_idx]} with total score {totals[best_idx]}")
    print(f"Worst graduating student is {names[worst_idx]} with total score {totals[worst_idx]}")
    
    class_total = sum(totals)
    class_avg = class_total / (num_students * num_subjects)
    
    print(f"Class total score: {class_total}")
    print(f"Class average score: {class_avg:.2f}")


if __name__ == "__main__":
    main()

File: javascript/studentGrade.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function askQuestion(question) {
  return new Promise(resolve => {
    rl.question(question, answer => {
      resolve(answer);
    });
  });
}

async function getPositiveInput(prompt) {
  let input;
  do {
    input = parseInt(await askQuestion(prompt));
    if (isNaN(input) || input <= 0) {
      console.log("Invalid input. Please enter a positive integer.");
    }
  } while (isNaN(input) || input <= 0);
  return input;
}

async function getValidScore(prompt, name, subject) {
  let score;
  do {
    score = parseInt(await askQuestion(`${prompt} for ${name} in subject ${subject} (0-100): `));
    if (isNaN(score) || score < 0 || score > 100) {
      console.log("Invalid score! Enter a value between 0-100.");
    }
  } while (isNaN(score) || score < 0 || score > 100);
  return score;
}

async function collectStudentData(numStudents, numSubjects) {
  const studentNames = [];
  const studentGrades = [];
  
  for (let i = 0; i < numStudents; i++) {
    const name = await askQuestion(`Enter name for student ${i+1}: `);
    studentNames.push(name);
    
    const grades = [];
    for (let j = 0; j < numSubjects; j++) {
      const score = await getValidScore("Enter score", name, j+1);
      grades.push(score);
    }
    studentGrades.push(grades);
  }
  
  return { studentNames, studentGrades };
}

function calculateTotals(studentGrades) {
  return studentGrades.map(grades => grades.reduce((sum, grade) => sum + grade, 0));
}

function calculateAverages(totals, numSubjects) {
  return totals.map(total => total / numSubjects);
}

function calculatePositions(totals) {
  const indexedTotals = totals.map((total, index) => ({ total, index }));
  indexedTotals.sort((a, b) => b.total - a.total);
  
  const positions = new Array(totals.length);
  let currentRank = 1;
  
  indexedTotals.forEach((item, i) => {
    if (i > 0 && item.total < indexedTotals[i-1].total) {
      currentRank = i + 1;
    }
    positions[item.index] = currentRank;
  });
  
  return positions;
}

function printReportHeader(numStudents, numSubjects) {
  console.log("\nClass Summary Report");
  console.log("====================");
  console.log(`Number of Students: ${numStudents}`);
  console.log(`Number of Subjects: ${numSubjects}\n`);
  
  let header = "Student Name\t\t";
  for (let i = 0; i < numSubjects; i++) {
    header += `Subject ${i+1}\t`;
  }
  header += "Total\tAverage\tPosition";
  console.log(header);
}

function printStudentResults(names, grades, totals, averages, positions) {
  names.forEach((name, i) => {
    let row = `${name}\t\t`;
    row += grades[i].join("\t") + "\t";
    row += `${totals[i]}\t${averages[i].toFixed(2)}\t${positions[i]}`;
    console.log(row);
  });
}

function printSubjectSummaries(names, grades) {
  console.log("\nSubject Summaries");
  console.log("===============");
  
  const numSubjects = grades[0].length;
  const numStudents = grades.length;
  
  for (let j = 0; j < numSubjects; j++) {
    let highest = -1;
    let lowest = 101;
    let highName = "";
    let lowName = "";
    let total = 0;
    let passes = 0;
    
    for (let i = 0; i < numStudents; i++) {
      const score = grades[i][j];
      total += score;
      if (score > highest) {
        highest = score;
        highName = names[i];
      }
      if (score < lowest) {
        lowest = score;
        lowName = names[i];
      }
      if (score >= 50) passes++;
    }
    
    const avg = total / numStudents;
    const fails = numStudents - passes;
    
    console.log(`\nSubject ${j+1}`);
    console.log(`Highest scoring student is: ${highName} (${highest})`);
    console.log(`Lowest scoring student is: ${lowName} (${lowest})`);
    console.log(`Total score: ${total}`);
    console.log(`Average score: ${avg.toFixed(2)}`);
    console.log(`Number of passes: ${passes}`);
    console.log(`Number of fails: ${fails}`);
    console.log("---------------------");
  }
}

function printClassStatistics(names, grades, totals) {
  console.log("\nClass Statistics");
  console.log("================");
  

  const numSubjects = grades[0].length;
  const numStudents = grades.length;
  
  let maxFails = -1;
  let minFails = numStudents + 1;
  let hardest = -1;
  let easiest = -1;
  
  for (let j = 0; j < numSubjects; j++) {
    let fails = 0;
    for (let i = 0; i < numStudents; i++) {
      if (grades[i][j] < 50) fails++;
    }
    
    if (fails > maxFails) {
      maxFails = fails;
      hardest = j;
    }
    if (fails < minFails) {
      minFails = fails;
      easiest = j;
    }
  }
  
  console.log(`The hardest subject is Subject ${hardest+1} with ${maxFails} failures`);
  console.log(`The easiest subject is Subject ${easiest+1} with ${numStudents-minFails} passes`);
  
  let highScore = -1;
  let lowScore = 101;
  let highName = "";
  let lowName = "";
  let highSubj = -1;
  let lowSubj = -1;
  
  for (let i = 0; i < numStudents; i++) {
    for (let j = 0; j < numSubjects; j++) {
      const score = grades[i][j];
      if (score > highScore) {
        highScore = score;
        highName = names[i];
        highSubj = j;
      }
      if (score < lowScore) {
        lowScore = score;
        lowName = names[i];
        lowSubj = j;
      }
    }
  }
  
  console.log(`The overall highest score is scored by ${highName} in Subject ${highSubj+1} (${highScore})`);
  console.log(`The overall lowest score is scored by ${lowName} in Subject ${lowSubj+1} (${lowScore})`);
 
  console.log("\nClass Summary");
  console.log("=============");
  
  const bestIdx = totals.indexOf(Math.max(...totals));
  const worstIdx = totals.indexOf(Math.min(...totals));
  
  console.log(`Best graduating student is ${names[bestIdx]} with total score ${totals[bestIdx]}`);
  console.log(`Worst graduating student is ${names[worstIdx]} with total score ${totals[worstIdx]}`);
  
  const classTotal = totals.reduce((sum, total) => sum + total, 0);
  const classAvg = classTotal / (numStudents * numSubjects);
  
  console.log(`Class total score: ${classTotal}`);
  console.log(`Class average score: ${classAvg.toFixed(2)}`);
}

async function main() {
  console.log("How many students do you have?");
  const numStudents = await getPositiveInput("Enter a positive integer: ");
  
  console.log("\nHow many subjects do they offer?");
  const numSubjects = await getPositiveInput("Enter a positive integer: ");
  console.log("Saved successfully\n");
  
  const { studentNames, studentGrades } = await collectStudentData(numStudents, numSubjects);
  const totals = calculateTotals(studentGrades);
  const averages = calculateAverages(totals, numSubjects);
  const positions = calculatePositions(totals);
  
  printReportHeader(numStudents, numSubjects);
  printStudentResults(studentNames, studentGrades, totals, averages, positions);
  printSubjectSummaries(studentNames, studentGrades);
  printClassStatistics(studentNames, studentGrades, totals);
  
  rl.close();
}

main().catch(err => {
  console.error("An error occurred:", err);
  rl.close();
});